src/protocols/UBX.hpp

```diff
@@ -0,0 +1,135 @@
+#ifndef GPS_UBLOX_PROTOCOLS_UBX_HPP
+#define GPS_UBLOX_PROTOCOLS_UBX_HPP
+
+#include <stdint.h>
+#include <stddef.h>
+
+#include <array>
+#include <vector>
+
+#include <gps_ublox/protocols/Parser.hpp>
+
+namespace gps_ublox
+{
+    namespace protocols {
+        class UBX : public Parser
+        {
+            private:
+                /** Computes the 2-byte checksum over the given buffer
+                 */
+                static std::array<uint8_t, 2> checksum(const uint8_t *buffer, const uint8_t *end);
+
+            public:
+                static const uint8_t SYNC_1;
+                static const uint8_t SYNC_2;
+                static const uint8_t SYNC_1_IDX;
+                static const uint8_t SYNC_2_IDX;
+                static const uint8_t MSG_CLASS_IDX;
+                static const uint8_t MSG_ID_IDX;
+                static const uint8_t LENGTH_LSB_IDX; // the payload-only length (LSB)
+                static const uint8_t LENGTH_MSB_IDX; // the payload-only length (MSB)
+                static const uint8_t PAYLOAD_IDX;
+                static const size_t FRAMING_SIZE_OVERHEAD;
+
+                /** The message class enumeration
+                 */
+                enum MsgClass {
+                    MSG_CLASS_ACK = 0x05,
+                    MSG_CLASS_CFG = 0x06
+                };
+
+                /** The message id enumeration
+                 */
+                enum MsgId {
+                    MSG_ID_NACK = 0x00,
+                    MSG_ID_ACK = 0x01,
+                    MSG_ID_VALSET = 0x8A
+                };
+
+                /** The layer where a configuration data is stored
+                 */
+                enum ConfigLayer {
+                    LAYER_RAM = 1,
+                    LAYER_BBR = 2,
+                    LAYER_FLASH = 4,
+                    LAYER_ALL = 7
+                };
+
+                /** The unique key id of a configuration value
+                 */
+                enum ConfigKeyId {
+                    I2C_ENABLED = 0x10510003,
+                    I2C_IN_UBX = 0x10710001,
```

Comment:
Hey. I noticed that the UBX/NMEA/RTCM ... is actually a constant offset after the base of `I2C_IN`. It's the same thing with the message period settings. There is a message base (e.g. NAV_LLH) and a constant offset to apply it to different outputs (e.g. +1 for UART1 and so on). I think it would make a lot more sense to leverage that to simplify the interface.

Reply:
I don't really understand what you mean (but i like the sound of "simplify the interface").

Reply:
Basically,

~~~
I2C_IN_UBX        = 0x10710000 + 0x00000000 + 0x00000001
                               ^^^ I2C         + ^^^ IN           + ^^^ UBX
SPI_OUT_NMEA = 0x10790000 + 0x00010000 + 0x00000002
                               ^^^ SPI         + ^^^ OUT       + ^^^ NMEA
~~~

So, one can easily define e.g. `setChannelConfiguration(InOutChannel, bool output, InOutProtocol protocol);` and avoid having a long long list of key IDs.

The same general pattern exists for message configuration, there's a base and then offsets for each channels:

~~~
CFG-MSGOUT-NMEA_ID_DTM_I2C        0x209100a6
CFG-MSGOUT-NMEA_ID_DTM_UART1  0x209100a7
CFG-MSGOUT-NMEA_ID_DTM_UART2  0x209100a8
CFG-MSGOUT-NMEA_ID_DTM_USB       0x209100a9
CFG-MSGOUT-NMEA_ID_DTM_SPI        0x209100aa
~~~

Reply:
Nice 👍